applications/StructuralMechanicsApplication/custom_elements/cr_beam_element_3D2N.cpp

```diff
@@ -1,1981 +1,2032 @@
-		const int dimension = 3;
-
-
-		array_1d<double, 3> v3 = ZeroVector(dimension);
-
-		double VectorNorm;
-		VectorNorm = MathUtils<double>::Norm(v1);
-		if (VectorNorm != 0) v1 /= VectorNorm;
-
-		VectorNorm = MathUtils<double>::Norm(v2);
-		if (VectorNorm != 0) v2 /= VectorNorm;
-
-		v3 = MathUtils<double>::CrossProduct(v2, v1);
-		VectorNorm = MathUtils<double>::Norm(v3);
-		if (VectorNorm != 0) v3 /= VectorNorm;
-
-
-		Matrix RotationMatrix = ZeroMatrix(dimension);
-		for (int i = 0; i < dimension; ++i) {
-			RotationMatrix(i, 0) = v1[i];
-			RotationMatrix(i, 1) = v2[i];
-			RotationMatrix(i, 2) = v3[i];
-		}
-
-		this->GetQuaternion() = Quaternion<double>::FromRotationMatrix(RotationMatrix);
-
-		KRATOS_CATCH("")
-	}
-
-	void Orientation::CalculateRotationMatrix(Matrix& R) {
-
-		KRATOS_TRY
-			if (R.size1() != 3 || R.size2() != 3) R.resize(3, 3, false);
-		const Quaternion<double> q = this->GetQuaternion();
-		q.ToRotationMatrix(R);
-		KRATOS_CATCH("")
-	}
-
-	void Orientation::CalculateBasisVectors(array_1d<double, 3>& v1,
-		array_1d<double, 3>& v2,
-		array_1d<double, 3>& v3) {
-
-		KRATOS_TRY
-			const Quaternion<double> q = this->GetQuaternion();
-		Matrix R = ZeroMatrix(3);
-		q.ToRotationMatrix(R);
-		if (v1.size() != 3) v1.resize(3, false);
-		if (v2.size() != 3) v2.resize(3, false);
-		if (v3.size() != 3) v3.resize(3, false);
-
-		for (int i = 0; i < 3; ++i) {
-			v1[i] = R(i, 0);
-			v2[i] = R(i, 1);
-			v3[i] = R(i, 2);
-		}
-		KRATOS_CATCH("")
-	}
-
-
-	void CrBeamElement3D2N::save(Serializer& rSerializer) const
-	{
-		KRATOS_SERIALIZE_SAVE_BASE_CLASS(rSerializer, Element);
-		rSerializer.save("DeformationModes", this->mDeformationModes);
-		rSerializer.save("NodalPosition", this->mTotalNodalPosistion);
-		rSerializer.save("NodalDeformation", this->mTotalNodalDeformation);
-		rSerializer.save("IterationCounter", this->mIterationCount);
-		rSerializer.save("NodalForces", this->mNodalForces);
-
-
-		rSerializer.save("LocalInitalAxisX", this->mNX0);
-		rSerializer.save("LocalInitalAxisY", this->mNY0);
-		rSerializer.save("LocalInitalAxisZ", this->mNZ0);
-
-
-		rSerializer.save("QuaternionVecA", this->mQuaternionVEC_A);
-		rSerializer.save("QuaternionVecB", this->mQuaternionVEC_B);
-		rSerializer.save("QuaternionScaA", this->mQuaternionSCA_A);
-		rSerializer.save("QuaternionScaB", this->mQuaternionSCA_B);
-
-		rSerializer.save("mIsLumpedMassMatrix", this->mIsLumpedMassMatrix);
-	}
-
-	void CrBeamElement3D2N::load(Serializer& rSerializer)
-	{
-		KRATOS_SERIALIZE_LOAD_BASE_CLASS(rSerializer, Element);
-		rSerializer.load("DeformationModes", this->mDeformationModes);
-		rSerializer.load("NodalPosition", this->mTotalNodalPosistion);
-		rSerializer.load("NodalDeformation", this->mTotalNodalDeformation);
-		rSerializer.load("IterationCounter", this->mIterationCount);
-		rSerializer.load("NodalForces", this->mNodalForces);
-
-		rSerializer.load("LocalInitalAxisX", this->mNX0);
-		rSerializer.lo
```

Comment:
Returning the Matrix by value may be a performance issue as I'm not sure if this matrix implements move semantic.